src/shims/unix/linux/epoll.rs

```diff
@@ -11,24 +13,56 @@ use self::shims::unix::fd::FileDescriptor;
 #[derive(Clone, Debug, Default)]
 struct Epoll {
     /// The file descriptors we are watching, and what we are watching for.
-    file_descriptors: FxHashMap<i32, EpollEvent>,
+    interest_list: BTreeMap<(WeakFileDescriptor, i32), Rc<EpollEvent>>,
+    // ready_list is an Rc because EpollEvents need to hold a reference to update
+    // it.
+    ready_list: Rc<RefCell<BTreeMap<(WeakFileDescriptor, i32), EpollReturn>>>,
+}
+
+/// EpollReturn contains information that will be returned by epoll_wait,
+/// and stored in ready list.
+#[derive(Debug)]
+pub struct EpollReturn {
+    // Events that happened to the file description.
+    events: u32,
+    // Original data retrieved from `epoll_event`
+    data: u64,
+}
+
+impl EpollReturn {
+    pub fn new(events: u32, data: u64) -> EpollReturn {
+        EpollReturn { events, data }
+    }
+    pub fn update_events(&mut self, flag: u32) {
+        self.events |= flag;
+    }
 }
 
 /// Epoll Events associate events with data.
-/// These fields are currently unused by miri.
-/// This matches the `epoll_event` struct defined
+/// `events` and `data` field matches the `epoll_event` struct defined
 /// by the epoll_ctl man page. For more information
 /// see the man page:
 ///
 /// <https://man7.org/linux/man-pages/man2/epoll_ctl.2.html>
 #[derive(Clone, Debug)]
-struct EpollEvent {
-    #[allow(dead_code)]
-    events: u32,
-    /// `Scalar` is used to represent the
-    /// `epoll_data` type union.
-    #[allow(dead_code)]
-    data: Scalar,
+pub struct EpollEvent {
+    // The file descriptor value associated with this epoll_event.
+    pub file_descriptor: i32,
+    // The file descriptor struct associated with this epoll_event.
+    pub weak_file_descriptor: WeakFileDescriptor,
```

Comment:
The terminology here doesn't quite work. `WeakFileDescriptor` points to the underlying *file description*. It is not a file descriptor.

Probably we should adjust the terminology we have in Miri already. I would suggest renaming the existing `FileDescriptor` to `FileDescriptionRef`, since arguably a FileDescriptor is indeed just an `i32`. Then `WeakFileDescriptor` would be `WeakFileDescriptionRef`.

Why do we even need both here in `EpollEvent`? Can we instead just store a `EpollKey` or so, or do we actually need to look at these fields?

Reply:
> Why do we even need both here in EpollEvent? Can we instead just store a EpollKey or so, or do we actually need to look at these fields?

Yup we need direct access to ``weak_file_description_ref`` for ``EPOLL_CTL_DEL`` as of now.